### PR TITLE
Streamline releasing and testing

### LINKED ISSUE #65: Configure coveralls.io
https://coveralls.io/r/ioam/param

Make sure it's accurately reporting coverage (see e.g. https://github.com/ioam/param/issues/14#issuecomment-43440149 and comment below it), then add a link/badge so we can easily see test coverage.

### LINKED ISSUE #86: nosetests should display warnings
E.g. there's at least one warning when the current test suite runs, but nobody's been aware of it until now (I just found it by chance):

```
root: WARNING: Parameterized00479: Setting non-parameter attribute print_level=0 using a mechanism intended only for parameters
```

### LINKED ISSUE #167: Clean up release procedure
https://github.com/ioam/param/wiki/Release-instructions is out of date, and contains code fragments that could potentially be added to the repository (e.g. conda package info). Also, current param conda packages contain various extraneous files and data.

### LINKED ISSUE #189: Travis: test different environments
* minimal environment: param is supposed to have no dependencies, but most pythons on travis seem to include numpy and a whole bunch of other stuff... 
```
asn1crypto==0.22.0
backports.shutil-get-terminal-size==1.0.0
certifi==2017.7.27.1
cffi==1.10.0
chardet==3.0.4
configparser==3.5.0
coverage==4.4.1
coveralls==1.2.0
cryptography==2.0.3
decorator==4.1.2
docopt==0.6.2
enum34==1.1.6
flake8==3.4.1
funcsigs==1.0.2
idna==2.6
ipaddress==1.0.18
ipython==5.3.0
ipython-genutils==0.2.0
mccabe==0.6.1
mock==2.0.0
nose==1.3.7
numpy==1.13.1
pathlib2==2.3.0
pbr==3.1.1
pexpect==4.2.1
pickleshare==0.7.4
prompt-toolkit==1.0.15
ptyprocess==0.5.2
py==1.4.34
pycodestyle==2.3.1
pycparser==2.18
pyflakes==1.5.0
Pygments==2.2.0
pyOpenSSL==17.2.0
pytest==3.2.1
requests==2.18.4
scandir==1.5
simplegeneric==0.8.1
six==1.10.0
traitlets==4.3.2
urllib3==1.22
wcwidth==0.1.7
```
(that's after installing coveralls, flake8, and ipython; https://travis-ci.org/ioam/param/jobs/273949060#L625)

-----

* numpy: although numpy isn't a dependency, I think we do various tests for it and enable certain features if numpy is available.

-----

* ipython: we have ipython extensions

-----

* anything else?

### LINKED ISSUE #192: Travis should test setup.py and resulting package

### LINKED ISSUE #195: Test on appveyor
To catch any code that's not cross platform

### PR DESCRIPTION
**Pull request by ceball**
This is a PR about streamlining param releases (fixes #167), and 
improving testing.

1. Added conda recipe and package upload

    Builds and uploads conda packages for (non-PR/simulated) commits to master.  I.e. latest master is always available as a conda package from anaconda.org/cball/param.

2. Added pypi package building and uploading

    For tags, builds sdist and universal wheel, and uploads to test.pypi (simulating actual releases to pypi).

3. Enabled testing on Windows via Appveyor.

    Param is pure python, but it's useful to test on windows to catch accidental platform specific content (whether that's code, developer workflow, testing, or whatever). Fixes #195.

4. Use tox for testing

    There is now very little scripting in the travis and appveyor config files (except for conda, which has no integration with travis). The same tests can be run locally as on CI. Nothing is platform specific.

    Tox not only tests 'python setup.py develop', but also creates an sdist, installs it, and tests it. Fixes #192.

    Changes include:

      * Moved dependencies out of platform-specific travis file and into setup.py (if genuine dependency) or tox (if test dependency).

      * Now tests that optional dependencies are actually optional. Previously, many other packages were present in the test environment. Have also started to test that specific functionality works when the optional dependencies are present. Fixes #189.

      * Fixed running of flake8.

      * Warnings during tests are now displayed (fixes #86).

    Being python only, param can take advantage of the standard python ecosystem, so unfortunately not all of this approach is suitable for conda-dependent packages.

Also, minor updates to README, including coverage badge (not sure why hiding it makes sense; closes #65).

### CODE REVIEW

**Comment by ceball** [comment]
Still got two things to address. 

## Can't build or test sdist between releases

sdist is the basis of a lot of packaging, and those packages are things that actually get installed, so it makes sense to test sdist. However, can't currently do that:

```
$ git status
On branch pkg
Your branch is ahead of 'origin/pkg' by 1 commit.
  (use "git push" to publish your local commits)
nothing to commit, working tree clean
$ git describe --dirty
v1.5.1-29-g400ab17
$ python setup.py sdist
Traceback (most recent call last):
  File "setup.py", line 45, in <module>
    param.__version__.verify(setup_args['version'])
  File "/Users/cball/code/ioam/param6/param/version.py", line 316, in verify
    raise Exception("Supplied string version does not match current version.")
Exception: Supplied string version does not match current version.
```

## "Latest master" conda packages will not have useful version info

Although the packages themselves will have a reasonable version (see https://github.com/ioam/parambokeh/pull/15), the installed param package will have a version of just e.g. `1.5.1`.

File: .appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+environment:
+  matrix:
+    - TOXENV: py36
+    - TOXENV: py35
+    - TOXENV: py34
+    - TOXENV: py33
+    - TOXENV: py27
+    - TOXENV: py26
+
```
Comment (jbednar) [inline .appveyor.yml]: I don't suppose anyone has written a command for Appveyor that will import the relevant bits from .travis.yml to avoid duplication?

File: .travis.yml [hunk 2]
```diff
@@ -1,27 +1,85 @@
...
+      env: TOX_ENV=py36
+      install:
+        - pip install tox
+      script:
+        - tox -e $TOX_ENV
+
+    - <<: *tests
+      python: 3.5
+      env: TOX_ENV=py35
+
```
Comment (jbednar) [inline .travis.yml]: It's a shame that this part had to get so verbose, but it seems worth it.

File: doc/index.rst [hunk 3]
```diff
@@ -135,19 +135,27 @@ Installation
...
+https://github.com/ioam/param/archive/master.zip``.
+
+For development, the `git repository <http://github.com/ioam/param>`_
+can be cloned and then 'develop installed' (``pip install -e .`` or
+``python setup.py develop``). Tests can be run via `tox
+<https://tox.readthedocs.io/en/latest/>`_: ``tox`` for all tests, or
+e.g. ``tox -e coverage`` to run unit tests with coverage for the
+currently active python. Alternatively, unit tests can be run via
+``nosetests`` (after installing `nose
+<http://nose.readthedocs.io/en/latest/>`_).
```
Comment (jbednar) [inline doc/index.rst]: So you don't want to switch from nose to pytest?

Comment (ceball) [reply doc/index.rst]: Yes, I probably do want to - if only to match the bulk of other projects I encounter right now, not because I understand the difference ;)

Was planning to do at the same time as parambokeh.

File: .appveyor.yml (hunk 1 above)
Comment (ceball) [reply .appveyor.yml]: Yep, sigh. At least it's overall a pretty short file now.

I think the ideal thing would be if appveyor and travis supported tox, then you'd only have the tox file. One problem on travis is that the 'build stages' I use here were added to travis fairly recently (they are still beta), so not even all of travis's own stuff seems to work with them yet. Otherwise I could maybe have used something like https://tox-travis.readthedocs.io/en/stable/.

I have used tox before on another project in the past (so it's been around a bit), and liked it. What I mainly like is that: 
 * the config is short...
 * ...yet it covers a lot (particularly the testing of packages)
 * and the config is separate from your code, i.e. does not tie itself in to your code. So (a) it is not a big deal to replace if something else comes along, and (b) I can use tox locally, but e.g. you don't have to if you don't want (you can just run one or more of the commands from tox.ini manually if you want, like `nosetests`).
 
It would be great if there were something providing similar functionality in the conda world, and if there were a conda plugin for travis, etc :)

**Comment by ceball** [comment]
> I'm not sure about the status of your two TODO items; looks like one can only be tested once on master 

Right

> but I'm not sure about the other one

This is the one I need to do something about (not just for this PR, but for parambokeh conda packages too). There are two parts to it.

First, I'd like the python package in the conda package to have a meaningful version. I'm considering something like this (proposed by Jean-Luc): `python -m param.version --set-conda-info .`, which could add `__version__.commit_count = <N>` and `__version__.commit = <SHA>` after the existing `__version__` in `__init__` (would need to change those being from read-only properties in param.version).

Second, I can't actually make sdist between tags at the moment, even to test them, without commenting out a version check. Jean-Luc has proposed modifying the commented-out guard in setup.py to `if ('upload' in sys.argv) or ('sdist' in sys.argv) and not 'PARAM_DISABLE_VERSION_CHECK' in os.environ:`, so I'm considering that. (E.g. I don't even know if 'upload' is a relevant argument or not any more for pypi uploading - need to check.)

File: setup.py [hunk 4]
```diff
@@ -40,9 +49,12 @@
...
 if __name__=="__main__":
 
-    if ('upload' in sys.argv) or ('sdist' in sys.argv):
-        import param, numbergen
-        param.__version__.verify(setup_args['version'])
-        numbergen.__version__.verify(setup_args['version'])
+# TODO: conda packages won't have meaningful version number yet
+
+# TODO: need to decide what to do about this
+#    if ('upload' in sys.argv) or ('sdist' in sys.argv):
```
Comment (jlstevens) [inline setup.py]: You can condition this on the presence of an environment variable when you need this check disabled.

Comment (ceball) [reply setup.py]: That's what my earlier comment says ;) It's hard to follow because this same discussion was happening in parambokeh too. I'll try to keep it here, and I'll ping you to review when I actually try something. Thanks!

(I've never quite understood this line, because aside from sdist there's also various bdist args one might pass instead. And nowadays I'm not sure anyone would do 'python setup.py upload' - I will check that. So some of the delay is that I need to update myself on python packaging.)

**Comment by ceball** [comment]
The recent hacks mean this can't currently be merged. But they show behavior that's getting closer to what I want. 

```
$ git describe --dirty
v1.5.1-27-gf7e059d
$ PARAM_MAKING_UNOFFICIAL_RELEASE=1 python setup.py sdist
[...]
creating param-1.5.1.dev27+gf7e059d
[...]
$ mv dist/param-1.5.1.dev27+gf7e059d.tar.gz /tmp/
$ pushd /tmp/
$ tar xf param-1.5.1.dev27+gf7e059d.tar.gz 
$ cd param-1.5.1.dev27+gf7e059d/
$ python
>>> import param
>>> param.__version__
1.5.1.dev27+gf7e059d
```

Also, `PARAM_MAKING_UNOFFICIAL_RELEASE=1 conda build conda.recipe/` gives me a conda package with the name `param-1.5.1.27_gf7e059d-pyh2898b3f_0.tar.bz2`; when I do `conda install -c (the dev channel) param` and then I import param, I'll get a version of `1.5.1.dev27+gf7e059d`.

**Comment by ceball** [comment]
@jlstevens and @jbednar: the recent commits to this PR leave it demonstrating (at least in my mind...) how versioneer allows the workflow described in https://github.com/ioam/autover/issues/2. (Albeit I have not changed from the 'pre' versioning scheme I described before Jim suggested a better alternative.)

However: the >2000 lines of versioneer I'm sure revokes Jim's previous review approval ;)

I don't propose to merge this, but I would love to have all projects operating like this. conda packages of master are always available, and 'release candidates' can be made any time with a `git tag ... && git push --tags`.

**Comment by ceball** [comment]
And https://github.com/ioam/autover/pull/9 sets up the same workflow on autover itself (but not working), as a target.

**Comment by jbednar** [comment]
Apart from the 2000 lines, that all sounds good to me.   Hopefully Jean-Luc can get that down to 20.

**Comment by ceball** [comment]
This PR's been open for over 2 months now. @jbednar, I've removed the versioneer demo, so do you re-approve of merging?

File: tests/testnumpy.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
+"""
+If numpy's present, is numpy stuff ok?
```
Comment (jlstevens) [inline tests/testnumpy.py]: Might be nice to have a better description here...

**Comment by jbednar** [comment]
Yes, I re-approve of merging.
